File: parakeet-rs/src/model_eou.rs
use crate::error::{Error, Result};
use crate::execution::ModelConfig as ExecutionConfig;
use ndarray::{Array1, Array2, Array3, Array4};
use ort::session::Session;
use std::path::Path;

/// Encoder cache state for streaming inference
/// The cache maintains temporal context across chunks
#[derive(Default)]
pub struct EncoderCache {
    /// channel cache: [1, 1, 70, 512] - batch=1, 70 frame lookback
    pub cache_last_channel: Array4<f32>,
    /// time cache: [1, 1, 512, 8] - batch=1, fixed 8 time steps
    pub cache_last_time: Array4<f32>,
    /// cache length: [1] with value 0 initially
    pub cache_last_channel_len: Array1<i64>,
}

impl EncoderCache {
    /// 17 layers, batch=1, 70 frame lookback, 512 features
    pub fn new() -> Self {
        Self {
            cache_last_channel: Array4::zeros((17, 1, 70, 512)),
            cache_last_time: Array4::zeros((17, 1, 512, 8)),
            cache_last_channel_len: Array1::from_vec(vec![0i64]),
        }
    }
}

pub struct ParakeetEOUModel {
    encoder: Session,
    decoder_joint: Session,
}

impl ParakeetEOUModel {
    pub fn from_pretrained<P: AsRef<Path>>(
        model_dir: P,
        exec_config: ExecutionConfig,
    ) -> Result<Self> {
        let model_dir = model_dir.as_ref();

        let encoder_path = model_dir.join("encoder.onnx");
        let decoder_path = model_dir.join("decoder_joint.onnx");

        if !encoder_path.exists() || !decoder_path.exists() {
            return Err(Error::Config(format!(
                "Missing ONNX files in {}. Expected encoder.onnx and decoder_joint.onnx",
                model_dir.display()
            )));
        }

        // Load encoder
        let builder = Session::builder()?;
        let builder = exec_config.apply_to_session_builder(builder)?;
        let encoder = builder.commit_from_file(&encoder_path)?;

        // Load decoder
        let builder = Session::builder()?;
        let builder = exec_config.apply_to_session_builder(builder)?;
        let decoder_joint = builder.commit_from_file(&decoder_path)?;

        Ok(Self {
            encoder,
            decoder_joint,
        })
    }

    /// Run the stateful encoder with cache
    /// Input: features [1, 128, T], cache state
    /// Output: (encoded [1, 512, T], new_cache)
    pub fn run_encoder(
        &mut self,
        features: &Array3<f32>,
        length: i64,
        cache: &EncoderCache,
    ) -> Result<(Array3<f32>, EncoderCache)> {
        let length_arr = Array1::from_vec(vec![length]);

        let outputs = self.encoder.run(ort::inputs![
            "audio_signal" => ort::value::Value::from_array(features.clone())?,
            "length" => ort::value::Value::from_array(length_arr)?,
            "cache_last_channel" => ort::value::Value::from_array(cache.cache_last_channel.clone())?,
            "cache_last_time" => ort::value::Value::from_array(cache.cache_last_time.clone())?,
            "cache_last_channel_len" => ort::value::Value::from_array(cache.cache_last_channel_len.clone())?
        ])?;

        // Extract encoder output [1, 512, T]
        let (shape, data) = outputs["outputs"]
            .try_extract_tensor::<f32>()
            .map_err(|e| Error::Model(format!("Failed to extract encoder output: {e}")))?;

        let shape_dims = shape.as_ref();
        let b = shape_dims[0] as usize;
        let d = shape_dims[1] as usize;
        let t = shape_dims[2] as usize;

        let encoder_out = Array3::from_shape_vec((b, d, t), data.to_vec())
            .map_err(|e| Error::Model(format!("Failed to reshape encoder output: {e}")))?;

        // Extract new cache states
        let (ch_shape, ch_data) = outputs["new_cache_last_channel"]
            .try_extract_tensor::<f32>()
            .map_err(|e| Error::Model(format!("Failed to extract cache_last_channel: {e}")))?;

        let (tm_shape, tm_data) = outputs["new_cache_last_time"]
            .try_extract_tensor::<f32>()
            .map_err(|e| Error::Model(format!("Failed to extract cache_last_time: {e}")))?;

        let (len_shape, len_data) = outputs["new_cache_last_channel_len"]
            .try_extract_tensor::<i64>()
            .map_err(|e| Error::Model(format!("Failed to extract cache_len: {e}")))?;

        // Build new cache with extracted shapes
        let new_cache = EncoderCache {
            cache_last_channel: Array4::from_shape_vec(
                (
                    ch_shape[0] as usize,
                    ch_shape[1] as usize,
                    ch_shape[2] as usize,
                    ch_shape[3] as usize,
                ),
                ch_data.to_vec(),
            )
            .map_err(|e| Error::Model(format!("Failed to reshape cache_last_channel: {e}")))?,

            cache_last_time: Array4::from_shape_vec(
                (
                    tm_shape[0] as usize,
                    tm_shape[1] as usize,
                    tm_shape[2] as usize,
                    tm_shape[3] as usize,
                ),
                tm_data.to_vec(),
            )
            .map_err(|e| Error::Model(format!("Failed to reshape cache_last_time: {e}")))?,

            cache_last_channel_len: Array1::from_shape_vec(
                len_shape[0] as usize,
                len_data.to_vec(),
            )
            .map_err(|e| Error::Model(format!("Failed to reshape cache_len: {e}")))?,
        };

        Ok((encoder_out, new_cache))
    }

    /// Run the stateful decoder
    /// Returns: (logits [1, 1, 1, vocab], new_state_h, new_state_c)
    pub fn run_decoder(
        &mut self,
        encoder_frame: &Array3<f32>, // [1, 512, 1]
        last_token: &Array2<i32>,    // [1, 1]
        state_h: &Array3<f32>,       // [1, 1, 640]
        state_c: &Array3<f32>,       // [1, 1, 640]
    ) -> Result<(Array3<f32>, Array3<f32>, Array3<f32>)> {
        // Target length is always 1 for single step
        let target_len = Array1::from_vec(vec![1i32]);

        let outputs = self.decoder_joint.run(ort::inputs![
            "encoder_outputs" => ort::value::Value::from_array(encoder_frame.clone())?,
            "targets" => ort::value::Value::from_array(last_token.clone())?,
            "target_length" => ort::value::Value::from_array(target_len)?,
            "input_states_1" => ort::value::Value::from_array(state_h.clone())?,
            "input_states_2" => ort::value::Value::from_array(state_c.clone())?
        ])?;

        // 1. Extract Logits
        let (l_shape, l_data) = outputs["outputs"]
            .try_extract_tensor::<f32>()
            .map_err(|e| Error::Model(format!("Failed to extract logits: {e}")))?;

        // 2. Extract States (output_states_1, output_states_2)
        let (_h_shape, h_data) = outputs["output_states_1"]
            .try_extract_tensor::<f32>()
            .map_err(|e| Error::Model(format!("Failed to extract state h: {e}")))?;

        let (_c_shape, c_data) = outputs["output_states_2"]
            .try_extract_tensor::<f32>()
            .map_err(|e| Error::Model(format!("Failed to extract state c: {e}")))?;

        // Reconstruct Arrays
        // Logits: I simplify to [1, 1, vocab]
        let vocab_size = l_shape[3] as usize;
        let logits = Array3::from_shape_vec((1, 1, vocab_size), l_data.to_vec())
            .map_err(|e| Error::Model(format!("Reshape logits failed: {e}")))?;

        // States: [1, 1, 640]
        let new_h = Array3::from_shape_vec((1, 1, 640), h_data.to_vec())
            .map_err(|e| Error::Model(format!("Reshape state h failed: {e}")))?;

        let new_c = Array3::from_shape_vec((1, 1, 640), c_data.to_vec())
            .map_err(|e| Error::Model(format!("Reshape state c failed: {e}")))?;

        Ok((logits, new_h, new_c))
    }
}

File: src/model_eou.py
import os
import numpy as np
import onnxruntime as ort
from dataclasses import dataclass, field
from numpy.typing import NDArray

class ModelError(Exception):
    pass

@dataclass
class EncoderCache:
    """
    Encoder state cache to maintain temporal context 
    across chunks during streaming inference.
    """
    # channel cache: [17, 1, 70, 512] - 17 layers, batch=1, 70 frame lookback
    cache_last_channel: NDArray = field(
        default_factory=lambda: np.zeros((17, 1, 70, 512), dtype=np.float32)
    )
    # time cache: [17, 1, 512, 8] - 17 layers, batch=1, fixed 8 time steps
    cache_last_time: NDArray = field(
        default_factory=lambda: np.zeros((17, 1, 512, 8), dtype=np.float32)
    )
    # cache length: [1]
    cache_last_channel_len: NDArray = field(
        default_factory=lambda: np.array([0], dtype=np.int64)
    )


class EOUModel:
    def __init__(
            self, 
            encoder_session: ort.InferenceSession, 
            decoder_session: ort.InferenceSession
            ):
        self.encoder = encoder_session
        self.decoder_joint = decoder_session

    @classmethod
    def from_pretrained(cls, model_dir: str, device: str = 'cpu') -> 'EOUModel':
        """
        Convenience method to load an EOU model consisting of an encoder 
        and a joint decoder from an ONNX model directory.

        Args:
            model_dir (str): Path to the model files.
            device (str): Device to use for inference, e.g. 'cpu' (default) or 'cuda'.

        Raises:
            ModelError: When any of the required model files cannot be found.

        Returns:
            EOUModel: Instance of an EOU model.
        """
        encoder_path = os.path.join(model_dir, "encoder_int8.onnx")
        decoder_path = os.path.join(model_dir, "decoder_joint.onnx")

        if not os.path.exists(encoder_path) or not os.path.exists(decoder_path):
            raise ModelError(
                f"Missing ONNX files in {model_dir}. Expected encoder.onnx and decoder_joint.onnx"
            )
        
        sess_options = ort.SessionOptions()
        sess_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL

        providers = ["CPUExecutionProvider"]
        if device.lower().startswith('cuda'):
            providers.insert(0, "CUDAExecutionProvider")

        encoder_session = ort.InferenceSession(
            path_or_bytes=encoder_path, 
            providers=providers,
            sess_options=sess_options
            )

        decoder_session = ort.InferenceSession(
            path_or_bytes=decoder_path, 
            providers=providers,
            sess_options=sess_options
            )

        return cls(encoder_session, decoder_session)

    def run_encoder(
        self,
        features: NDArray,
        length: int,
        cache: EncoderCache
        ) -> tuple[NDArray, EncoderCache]:
        """Runs the stateful encoder with cache.

        Args:
            features (NDArray): Input audio features with shape [1, 128, T].
            length (int): Length of the input sequence.
            cache (EncoderCache): Encoder cache from the previous step.

        Returns:
            Tuple[NDArray, EncoderCache]:
                - Encoded features with shape [1, 512, T].
                - Updated encoder cache.
        """

        outputs = self.encoder.run(
            output_names=None,
            input_feed={
                "audio_signal": features.astype(np.float32, copy=False),
                "length": np.asarray([length], dtype=np.int64),
                "cache_last_channel": cache.cache_last_channel,
                "cache_last_time": cache.cache_last_time,
                "cache_last_channel_len": cache.cache_last_channel_len.astype(np.int64, copy=False),
            },
        )
        encoder_out, _, last_channel, last_time, last_channel_len = outputs

        cache = EncoderCache(
            cache_last_channel=last_channel,
            cache_last_time=last_time,
            cache_last_channel_len=last_channel_len,
        )

        return encoder_out, cache

    def run_decoder(
        self,
        encoder_frame: NDArray,
        last_token: NDArray,
        state_h: NDArray,
        state_c: NDArray,
    ) -> tuple[NDArray, NDArray, NDArray]:
        """
        Runs the stateful decoder for a single step.

        Args:
            encoder_frame (NDArray): Encoder output frame with shape [1, 512, 1].
            last_token (NDArray): Previous token ID with shape [1, 1].
            state_h (NDArray): Previous decoder hidden state with shape [1, 1, 640].
            state_c (NDArray): Previous decoder cell state with shape [1, 1, 640].

        Returns:
            Tuple[NDArray, NDArray, NDArray]:
                - Logits with shape [1, 1, vocab].
                - Updated hidden state with shape [1, 1, 640].
                - Updated cell state with shape [1, 1, 640].
        """
        outputs = self.decoder_joint.run(
            output_names=None,
            input_feed={
                "encoder_outputs": encoder_frame.astype(np.float32, copy=False),
                "targets": last_token.astype(np.int32, copy=False),
                "target_length": np.array([1], dtype=np.int32),
                "input_states_1": state_h.astype(np.float32, copy=False),
                "input_states_2": state_c.astype(np.float32, copy=False),
            },
        )
        logits, _, new_h, new_c = outputs

        return logits, new_h, new_c
